File: Notes/CTCI/01-arrays-strings/05.js
/* 1.5 - One Away
 *
 * There are three types of edits that can be performed on strings:
 * insert a character, remove a character, or replace a character.
 *
 * Given two strings, write a function to check if they are one edit 
 * (or zero edits) away.
 *
 * Example
 * pale, ple -> true
 * pales, pale -> true
 * pale, bale -> true
 * pale, bake -> false
*/

const oneAway = (s1, s2) => {
  
  if (Math.abs(s1.length - s2.length) > 1) return false;

  const s1Map = {};
  let missing = false;

  for (let i = 0; i < s1.length; i++) {
    s1Map[s1[i]] = (s1Map[s1[i]] ? s1Map[s1[i]]++ : 1);
  }
  
  for (let j = 0; j < s2.length; j++) {
    if (s2[j] in s1Map) {
      s1Map[s2[j]]--;

      if (s1Map[s2[j]] < -1) {
        return false;
      }

    } 
    else {
      if (missing) return false;
      missing = true;
    }
  }
  return true;
}

const CTCI = (s1, s2) => {
  // Length check
  if (Math.abs(s1.length - s2.length) > 1) return false;

  // Get shorter and longer strings
  let string1 = s1.length < s2.length ? s1 : s2;
  let string2 = s1.length < s2.length ? s2 : s1;

  let index1 = 0;
  let index2 = 0;

  let diff = false;

  while (index2 < s2.length && index1 < s1.length) {
    if (s1[index1] != s2[index2]) {

      if (diff) return false;
      diff = true;

      if (s1.length == s2.length) {
        index1++; // On replace, move shorter pointer
      }
    }
    else {
      index1++; // If matching, move shorter pointer
    }
    index2++; // Always move pointer for longer string
  }
  return true;
}

if (require.main === module) {
  console.log(oneAway(process.argv[2], process.argv[3]));
}

File: Notes/CTCI/01-arrays-strings/05.py
# 05. One Away
# Three types of edits: insert, remove, or replace. Given two strings, check if
# they are one (or zero) edits away.


def one_away(s1, s2):
    length_s1 = len(s1)
    length_s2 = len(s2)
    one_off = False
    j = 0

    if length_s1 > length_s2:
        if (length_s1 - length_s2 > 1):
            return False
        for i in range(length_s1):

            if i == length_s2:
                return True

            if s1[i] != s2[j]:
                if one_off:
                    return False
                one_off = True
                continue

            j += 1

        return True

    if length_s1 == length_s2:
        for i in range(length_s1):
            if s1[i] != s2[j]:
                if one_off:
                    return False
                one_off = True
            j += 1

        return True

    if length_s1 < length_s2:
        return one_away(s2, s1)


assert(one_away('pale', 'ple') == True)  # noqa: E721
assert(one_away('pales', 'pale') == True)  # noqa: E721
assert(one_away('pale', 'bale') == True)  # noqa: E721
assert(one_away('pale', 'bake') == False)  # noqa: E721
print('All tests passed!')
